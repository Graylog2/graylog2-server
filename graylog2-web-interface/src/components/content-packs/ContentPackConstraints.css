.success {
    background: #8DC63F;
}

.failure {
    background: #a94442;
}
