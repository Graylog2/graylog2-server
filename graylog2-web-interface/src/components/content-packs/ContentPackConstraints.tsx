/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */

import React, { useMemo, useCallback } from 'react';
import { Set } from 'immutable';

import { DataTable, StatusIcon } from 'components/common';
import useProductName from 'brand-customization/useProductName';

type ContentPackConstraintsProps = {
  constraints?: any | any[];
  isFulfilled?: boolean;
};

const ContentPackConstraints = ({ constraints = Set(), isFulfilled = false }: ContentPackConstraintsProps) => {
  const productName = useProductName();
  const headers = useMemo(() => ['Name', 'Type', 'Version', 'Fulfilled'], []);

  const renderHeader = useCallback((header: React.ReactNode) => <th>{header}</th>, []);

  const formattedConstraints = useMemo(() => {
    let updatedConstraints = constraints.map((constraint) => {
      const newConstraint = constraint.constraint || constraint;
      newConstraint.fulfilled = constraint.fulfilled;

      return newConstraint;
    });

    if (typeof updatedConstraints.toArray === 'function') {
      updatedConstraints = updatedConstraints.toArray();
    }

    return updatedConstraints;
  }, [constraints]);

  const rowFormatter = (item) => {
    const constraint = item.constraint || item;
    constraint.fulfilled = isFulfilled || constraint.fulfilled;
    const name = constraint.type === 'server-version' ? productName : constraint.plugin;

    return (
      <tr key={constraint.id}>
        <td>{name}</td>
        <td>{constraint.type}</td>
        <td>{constraint.version}</td>
        <td>
          <StatusIcon active={constraint.fulfilled} />
        </td>
      </tr>
    );
  };

  return (
    <div>
      <h2>Constraints</h2>
      <br />
      <br />
      <DataTable
        id="content-packs-constraints"
        headers={headers}
        headerCellFormatter={renderHeader}
        sortBy={(row) => (row.constraint ? row.constraint.type : row.type)}
        dataRowFormatter={rowFormatter}
        rows={formattedConstraints}
        filterKeys={[]}
      />
    </div>
  );
};

export default ContentPackConstraints;
