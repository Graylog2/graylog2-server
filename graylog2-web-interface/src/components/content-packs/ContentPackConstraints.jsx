import PropTypes from 'prop-types';
import React from 'react';

import { DataTable } from 'components/common';
import { Badge } from 'react-bootstrap';
import './ContentPackConstraints.css';

class ContentPackConstraints extends React.Component {
  static propTypes = {
    constraints: PropTypes.array,
    isFulfilled: PropTypes.bool,
  };

  static defaultProps = {
    constraints: [],
    isFulfilled: false,
  };

  _rowFormatter = (item) => {
    const constraint = item.constraint || item;
    const fulfilledIcon = item.fulfilled || this.props.isFulfilled ? <i className="fa fa-check" /> : <i className="fa fa-times" />;
    const fulfilledBg = item.fulfilled || this.props.isFulfilled ? 'success' : 'failure';
    return (
      <tr key={constraint.id}>
        <td>{constraint.type}</td>
        <td>{constraint.version}</td>
        <td><Badge className={fulfilledBg}>{fulfilledIcon}</Badge></td>
      </tr>
    );
  };

  render() {
    const headers = ['Type', 'Version', 'Fulfilled'];
    return (
      <div>
        <h2>Constraints</h2>
        <br />
        <br />
        <DataTable
          id="content-packs-constraints"
          headers={headers}
          headerCellFormatter={header => <th>{header}</th>}
          sortByKey="type"
          dataRowFormatter={this._rowFormatter}
          rows={this.props.constraints}
          filterKeys={[]}
        />
      </div>
    );
  }
}

export default ContentPackConstraints;
