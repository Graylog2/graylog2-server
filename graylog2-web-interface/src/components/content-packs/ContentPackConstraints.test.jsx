import React from 'react';
import { mount } from 'wrappedEnzyme';
import 'helpers/mocking/react-dom_mock';

import ContentPackConstraints from 'components/content-packs/ContentPackConstraints';

describe('<ContentPackConstraints />', () => {
  it('should render with new constraints without forced fulfillment', () => {
    const constraints = [{
      type: 'server-version',
      version: '>=3.0.0-alpha.2+af8d8e0',
    }, {
      plugin: 'org.graylog.plugins.threatintel.ThreatIntelPlugin',
      type: 'plugin-version',
      version: '>=3.0.0-alpha.2',
    }];
    const wrapper = mount(<ContentPackConstraints constraints={constraints} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with new constraints with forced fulfillment', () => {
    const constraints = [{
      type: 'server-version',
      version: '>=3.0.0-alpha.2+af8d8e0',
    }, {
      plugin: 'org.graylog.plugins.threatintel.ThreatIntelPlugin',
      type: 'plugin-version',
      version: '>=3.0.0-alpha.2',
    }];
    const wrapper = mount(<ContentPackConstraints constraints={constraints} isFulfilled />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with created constraints', () => {
    const constraints = [
      {
        constraint: {
          type: 'server-version',
          version: '>=3.0.0-alpha.2+af8d8e0',
        },
        fulfilled: true,
      }, {
        constraint: {
          plugin: 'org.graylog.plugins.threatintel.ThreatIntelPlugin',
          type: 'plugin-version',
          version: '>=3.0.0-alpha.2',
        },
        fulfilled: false,
      }];
    const wrapper = mount(<ContentPackConstraints constraints={constraints} />);
    expect(wrapper).toMatchSnapshot();
  });
});
