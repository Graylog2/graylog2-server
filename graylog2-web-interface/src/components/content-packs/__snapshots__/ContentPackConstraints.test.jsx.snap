// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentPackConstraints /> should render with created constraints 1`] = `
<div>
  <h2>
    Constraints
  </h2>
  <br />
  <br />
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="content-packs-constraints"
        >
          <table
            className="table undefined"
          >
            <thead>
              <tr>
                <th>
                  Name
                </th>
                <th>
                  Type
                </th>
                <th>
                  Version
                </th>
                <th>
                  Fulfilled
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  org.graylog.plugins.threatintel.ThreatIntelPlugin
                </td>
                <td>
                  plugin-version
                </td>
                <td>
                  &gt;=3.0.0-alpha.2
                </td>
                <td>
                  <span
                    className="failure badge"
                  >
                    <i
                      className="fa fa-times"
                    />
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Graylog
                </td>
                <td>
                  server-version
                </td>
                <td>
                  &gt;=3.0.0-alpha.2+af8d8e0
                </td>
                <td>
                  <span
                    className="success badge"
                  >
                    <i
                      className="fa fa-check"
                    />
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackConstraints /> should render with new constraints with forced fulfillment 1`] = `
<div>
  <h2>
    Constraints
  </h2>
  <br />
  <br />
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="content-packs-constraints"
        >
          <table
            className="table undefined"
          >
            <thead>
              <tr>
                <th>
                  Name
                </th>
                <th>
                  Type
                </th>
                <th>
                  Version
                </th>
                <th>
                  Fulfilled
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  org.graylog.plugins.threatintel.ThreatIntelPlugin
                </td>
                <td>
                  plugin-version
                </td>
                <td>
                  &gt;=3.0.0-alpha.2
                </td>
                <td>
                  <span
                    className="success badge"
                  >
                    <i
                      className="fa fa-check"
                    />
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Graylog
                </td>
                <td>
                  server-version
                </td>
                <td>
                  &gt;=3.0.0-alpha.2+af8d8e0
                </td>
                <td>
                  <span
                    className="success badge"
                  >
                    <i
                      className="fa fa-check"
                    />
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackConstraints /> should render with new constraints without forced fulfillment 1`] = `
<div>
  <h2>
    Constraints
  </h2>
  <br />
  <br />
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="content-packs-constraints"
        >
          <table
            className="table undefined"
          >
            <thead>
              <tr>
                <th>
                  Name
                </th>
                <th>
                  Type
                </th>
                <th>
                  Version
                </th>
                <th>
                  Fulfilled
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  org.graylog.plugins.threatintel.ThreatIntelPlugin
                </td>
                <td>
                  plugin-version
                </td>
                <td>
                  &gt;=3.0.0-alpha.2
                </td>
                <td>
                  <span
                    className="failure badge"
                  >
                    <i
                      className="fa fa-times"
                    />
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Graylog
                </td>
                <td>
                  server-version
                </td>
                <td>
                  &gt;=3.0.0-alpha.2+af8d8e0
                </td>
                <td>
                  <span
                    className="failure badge"
                  >
                    <i
                      className="fa fa-times"
                    />
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
